macos/mac_os_resource_snd.ksy

```diff
@@ -0,0 +1,280 @@
+meta:
+  id: mac_os_resource_snd
+  title: Classic MacOS Sound Resource
+  application: Sound Manager
+  endian: be
+  license: MIT
+  xref:
+    wikidata: Q7564684
+    mac-os-resource-type: 'snd '
+doc-ref: "https://developer.apple.com/library/archive/documentation/mac/pdf/Sound/Sound_Manager.pdf"
+doc: |
+  Sound resources were introduced in Classic MacOS with the Sound Manager program.
+  They can contain sound commands to generate sounds with given frequencies as well as sampled sound data.
+  They are mostly found in resource forks, but can occasionally appear standalone or embedded in other files.
+seq:
+  - id: format
+    type: u2
+  - id: num_data_formats
+    -orig-id: number_of_data_formats
+    type: u2
+    if: format == 1
+  - id: data_formats
+    type: data_format
+    repeat: expr
+    repeat-expr: num_data_formats
+    if: format == 1
+  - id: reference_count
+    type: u2
+    if: format == 2
+  - id: num_sound_commands
+    -orig-id: number_of_sound_commands
+    type: u2
+  - id: sound_commands
+    type: sound_command
+    repeat: expr
+    repeat-expr: num_sound_commands
+types:
+  data_format:
+    seq:
+      - id: id
+        type: u2
+        enum: data_type
+      - id: options
+        type: u4
+  sound_command:
+    seq:
+      - id: is_data_offset
+        type: b1
+      - id: cmd
+        type: b15
+        enum: cmd_type
+      - id: param1
+        type: u2
+      - id: param2
+        type: u4
+    instances:
+      sound_header_type:
+        pos: param2 + 20
+        type: u1
+        enum: sound_header_type
+        if: is_data_offset and cmd == cmd_type::buffer_cmd
+      sound_header:
+        pos: param2
+        size-eos: true
+        if: is_data_offset and cmd == cmd_type::buffer_cmd
+        type:
+          switch-on: sound_header_type
+          cases:
+            sound_header_type::standard: standard_sound_header
+            sound_header_type::extended: extended_sound_header
+            sound_header_type::compressed: compressed_sound_header
+  standard_sound_header:
+    seq:
+      - id: sample_ptr
+        -orig-id: samplePtr
+        type: u4
+        doc: pointer to samples (or 0 if samples follow data structure)
+      - id: length
+        type: u4
+        doc: number of samples in array
+      - id: sample_rate
+        -orig-id: sampleRate
+        type: unsigned_fixed_point
+        doc: The rate at which the sample was originally recorded.
+      - id: loop_start
+        -orig-id: loopStart
+        type: u4
+        doc: loop point beginning
+      - id: loop_end
+        -orig-id: loopEnd
+        type: u4
+        doc: loop point ending
+      - id: encode
+        type: u1
+        enum: sound_header_type
+        doc: sample's encoding option
+      - id: base_frequency
+        type: u1
+        doc: base frequency of sample
+        -orig-id: baseFrequency
```

Comment:
It should be stated here in what units is the _frequency_ encoded here, or preferably directly create a value instance with a formula that converts this value to hertz (Hz), if it's possible.

Also if the value stored in this attribute is not directly the _frequency in Hertz_, it should not be called `base_frequency`.

Reply:
the frequency is represented as midi note. (60 being "middle C")

https://en.wikipedia.org/wiki/MIDI_tuning_standard

It can be converted to a frequency in hertz, but that needs power function support for floats and I don't think that is currently supported in Kaitai, or is it?

![MIDI to freqeuncy formula](https://wikimedia.org/api/rest_v1/media/math/render/svg/18487bf6ef8bd35b6e0c1af03f7b6381c8a8aa3f)

values from 0-127 are supported, so I could add a 127 entry lookup switch, but don't know if that is really wanted. 😄

Reply:
> values from 0-127 are supported, so I could add a 127 entry lookup switch, but don't know if that is really wanted. 😄

OK, you don't have to do that if you don't want to, but it actually wouldn't be so bad idea. Lookup tables can be defined in KS, and there's nothing wrong about that: in fact, it will be naturally faster than evaluating lots of expressions from scratch. And a one-time 1k bytes of memory (`8 * 128` for `double`) are nowadays a ridiculously small amount, so that isn't an issue either.

Neither does it take up much space in the spec: I wrote a JavaScript snippet putting 8 numbers per line with 2 decimals, and it would be 16 lines with max 70 chars per line, which is acceptable _(open DevTools with `F12` or `Ctrl+Shift+I` and paste, or maybe https://jsbin.com/ if you're on mobile)_:

```javascript
const lookupTable = [];
for (let d = 0; d < 127; d++) {
  lookupTable[d] = (2 ** ((d - 69) / 12)) * 440;
}

function printTable(table, {decimals, numsPerLine}) {
  return table
    .map(f => f.toFixed(decimals))
    .reduce((acc, val) => {
        let row = acc[acc.length - 1];
        if (row.length >= numsPerLine) {
          row = [];
          acc.push(row);
        }
        row.push(val);
        return acc;
    }, [[]])
    .map(a => a.join(", "))
    .join("\n");
}

console.log(printTable(lookupTable, {decimals: 2, numsPerLine: 8}));
```

Check out the MacOS dcmp_2 spec by @dgelessus for comparison 😁:

https://github.com/kaitai-io/kaitai_struct_formats/blob/d1e58e36f26ecff1375c6022dc040342914ee4d1/macos/resource_compression/dcmp_2.ksy#L89-L156

Reply:
@tisoft If you think that the lookup table is a bit overkill, then please at least put the formula `(2 ** ((d - 69) / 12)) * 440` in the `doc` key, that might be useful for the end user.

Reply:
Implemented the lookup table for the standard header, will unify the headers, so that it is available in all.